Confirm delta

clemahieu opened the pull request:
online_weight_minimum is already considered in the calculation of node::delta. Using online_weight_minimum directly can cause quorum requirements to be unnecessarily high as actual online vote weight approaches online_weight_minimum.

Replacing references to online_weight_minimum with node:delta and adding tests to ensure rollbacks and confirmation is done precisely on the correct conditions.

clemahieu committed:
Rewrite tests in terms of state blocks.

clemahieu committed:
online_weight_minimum is already considered in the calculation of node::delta. Using online_weight_minimum directly can cause quorum requirements to be unnecessarily high as the online_weight_minimum is approached.

Replacing references to online_weight_minimum with node:delta and adding tests to ensure rollbacks and confirmation is done precisely on the correct conditions.

clemahieu committed:
online_weight_minimum is already considered in the calculation of node::delta. Using online_weight_minimum directly can cause quorum requirements to be unnecessarily high as actual online vote weight approaches online_weight_minimum.

Replacing references to online_weight_minimum with node:delta and adding tests to ensure rollbacks and confirmation is done precisely on the correct conditions.

guilhermelawless commented:
Spots that might also need a change (or at least can be optimized) as part of this:
- `peers_stake_required` in the `confirmation_quorum` RPC response
- `gap_cache::bootstrap_check` could possibly use this threshold and not online weight minimum
- Similarly for `active_transactions::inactive_votes_bootstrap_check` , for legacy bootstrap when lazy is disabled

clemahieu committed:
Replace additional direct references to online_weight_minimum with references to delta (), which already considers the online_weight_minimum.

Srayman commented:
This handles the case where online weight is reduced and approaches the minimum, but what happens in the opposite scenario where a large rep comes back online or a large pending balance is received?  The peers_stake would increase immediately but online_weight is trended over 2 weeks so there would be a lag before delta is updated to account for the new weight.

Should delta be the greater of 50% of online weight or 50% of peers stake to account for the rapid increase in peers weight?  As the weight is reduced delta would slowly go down, but it would respond more dramatically to a rapid increase in weight since peers weight is real-time.

Alternatively modify the online weight process to move up quickly if the weight is increased but down slowly over 2 weeks if it decreases.

clemahieu committed:
Changing vote_processor to reference nano::node instead of all of its components directly.

clemahieu committed:
Moving online_stake calculation on to node.

clemahieu committed:
Tracking trended weight separately from currently online weight.
Moving delta on to online_reps class so all network weight calculations are done in one place.

clemahieu committed:
Merge branch 'develop' into confirm_delta

# Conflicts:
#	nano/core_test/confirmation_height.cpp
#	nano/node/active_transactions.cpp

clemahieu commented:
> This handles the case where online weight is reduced and approaches the minimum, but what happens in the opposite scenario where a large rep comes back online or a large pending balance is received? The peers_stake would increase immediately but online_weight is trended over 2 weeks so there would be a lag before delta is updated to account for the new weight.
> 
> Should delta be the greater of 50% of online weight or 50% of peers stake to account for the rapid increase in peers weight? As the weight is reduced delta would slowly go down, but it would respond more dramatically to a rapid increase in weight since peers weight is real-time.
> 
> Alternatively modify the online weight process to move up quickly if the weight is increased but down slowly over 2 weeks if it decreases.

@Srayman  Added some new changes to address this issue.

guilhermelawless commented on nano/core_test/election.cpp:
```diff
@@ -13,3 +13,155 @@ TEST (election, construction)
+	                                     .link (key1.pub)
+	                                     .work (0)
+	                                     .sign (nano::dev_genesis_key.prv, nano::dev_genesis_key.pub)
+	                                     .build ();
```
Can use `.build_shared()` on these which allows replacing `std::shared_ptr<nano::block>` with `auto`

guilhermelawless commented on nano/node/vote_processor.hpp:
```diff
@@ -33,7 +33,7 @@ namespace transport
 {
 public:
-	explicit vote_processor (nano::signature_checker & checker_a, nano::active_transactions & active_a, nano::node_observers & observers_a, nano::stat & stats_a, nano::node_config & config_a, nano::node_flags & flags_a, nano::logger_mt & logger_a, nano::online_reps & online_reps_a, nano::ledger & ledger_a, nano::network_params & network_params_a);
+	explicit vote_processor (nano::node & node_a);
```
Many class forward declarations can be removed in this file, and `node` should be declared. Does this change to reference `node` go against our efforts to not have `node` be referenced by other classes to reduce compilation time?

clemahieu replied:
I think this was some cruft left over from building the PR. At any rate it's an unrelated change so removed.

clemahieu committed:
Merge branch 'develop' into confirm_delta

# Conflicts:
#	nano/node/election.hpp
#	nano/node/online_reps.cpp
#	nano/node/online_reps.hpp

clemahieu committed:
Merge branch 'develop' into confirm_delta

# Conflicts:
#	nano/node/election.hpp
#	nano/node/vote_processor.cpp

clemahieu committed:
Cleaning up tests and removing unrelated vote_processor changes.

wezrule commented on nano/node/online_reps.cpp:
```diff
@@ -1,101 +1,131 @@
 	std::vector<nano::uint128_t> items;
-	items.reserve (network_params.node.max_weight_samples + 1);
-	items.push_back (minimum);
+	items.push_back (config.online_weight_minimum.number ());
+	items.reserve (config.network_params.node.max_weight_samples + 1);
```
Most implementations tend to give `std::vector` a default size of 0, so could reduce an allocation swapping these lines back (i.e `reserve` first)

wezrule commented on nano/node/vote_processor.cpp:
```diff
@@ -15,16 +15,24 @@
-stopped{ false },
-is_active{ false },
-thread{ [this]() {
+nano::vote_processor::vote_processor (nano::signature_checker & checker_a, nano::active_transactions & active_a, nano::node_observers & observers_a, nano::stat & stats_a, nano::node_config & config_a, nano::node_flags & flags_a, nano::logger_mt & logger_a, nano::online_reps & online_reps_a, nano::ledger & ledger_a, nano::network_params & network_params_a) :
```
Nice job removing the `node` dependency! Should be able to remove the `node.hpp` include now too

wezrule commented on nano/node/online_reps.cpp:
```diff
@@ -1,101 +1,131 @@
 	if (ledger.weight (rep_a) > 0)
 	{
+		auto now = std::chrono::steady_clock::now ();
 		nano::lock_guard<std::mutex> lock (mutex);
-		reps.insert (rep_a);
+		auto new_insert = reps.get<tag_account> ().erase (rep_a) == 0;
+		reps.insert ({ now, rep_a });
```
Should `now` be captured inside the critical section in case it takes a while to grab the mutex?

SergiySW replied:
Is emplace () better here?

clemahieu replied:
Emplace can't be used with POD type initialization.

SergiySW commented on nano/node/online_reps.cpp:
```diff
@@ -1,101 +1,131 @@
-	// Discard oldest entries
-	while (ledger.store.online_weight_count (transaction) >= network_params.node.max_weight_samples)
+	nano::unique_lock<std::mutex> lock (mutex);
+	nano::uint128_t online_l;
```
Can be `nano::uint128_t online_l (online_m);` or `nano::uint128_t online_l = online_m;` probably

clemahieu committed:
Avoiding extra memory allocations on vector.
Generating timestamp within mutex to avoid issues of delays.
Removing reference to unused header.
